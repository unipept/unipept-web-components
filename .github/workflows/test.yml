name: Test

on: [push]

jobs:
  lint:
    runs-on: ubuntu-latest
    
    steps:
    - uses: actions/checkout@v1
    - uses: actions/setup-node@v1
    - name: Install dependencies
      run: npm install
    - name: Install Jest CLI
      run: npm i -g jest-cli
    - name: Run tests
      run: jest
    
