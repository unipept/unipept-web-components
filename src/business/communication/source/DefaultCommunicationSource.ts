import CommunicationSource from "./CommunicationSource";
import EcResponseCommunicator from "@/business/communication/functional/ec/EcResponseCommunicator";
import GoResponseCommunicator from "./../functional/go/GoResponseCommunicator";
import InterproResponseCommunicator from "./../functional/interpro/InterproResponseCommunicator";
import NcbiResponseCommunicator from "./../taxonomic/ncbi/NcbiResponseCommunicator";
import Pept2DataCommunicator from "./../peptides/Pept2DataCommunicator";

export default class DefaultCommunicationSource implements CommunicationSource {
    public constructor(
        private readonly serviceUrl: string
    ) {}

    public getEcCommunicator(): EcResponseCommunicator {
        return new EcResponseCommunicator();
    }

    public getGoCommunicator(): GoResponseCommunicator {
        return new GoResponseCommunicator();
    }

    public getInterproCommunicator(): InterproResponseCommunicator {
        return new InterproResponseCommunicator();
    }

    public getNcbiCommunicator(): NcbiResponseCommunicator {
        return new NcbiResponseCommunicator();
    }

    public getPept2DataCommunicator(): Pept2DataCommunicator {
        return new Pept2DataCommunicator(this.serviceUrl);
    }
}
