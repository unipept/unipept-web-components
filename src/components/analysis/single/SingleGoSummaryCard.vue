<docs>
    A variant of the GoSummaryCard that's specifically designed for the analysis of a single peptide. This variant works
    with a peptide as input and provides the counted GO-terms in function of the amount of proteins associated with it.
</docs>

<template>
    <go-summary-card :loading="peptide === ''">
        <template v-slot:analysis-header>
            <span v-html="trustLine" class="go-trust"></span>
        </template>
        <template v-slot:content-biological-process>
            <go-summary
                :go-count-table="items[0].countTable"
                :namespace="namespaces[0]"
                :go-ontology="items[0].ontology"
                :relative-counts="trust ? trust.totalAmountOfItems : 1"
                :loading="loading"
                :show-percentage="false">
            </go-summary>
        </template>
        <template v-slot:content-cellular-component>
            <go-summary
                :go-count-table="items[1].countTable"
                :namespace="namespaces[1]"
                :go-ontology="items[1].ontology"
                :relative-counts="trust ? trust.totalAmountOfItems : 1"
                :loading="loading"
                :show-percentage="false">
            </go-summary>
        </template>
        <template v-slot:content-molecular-function>
            <go-summary
                :go-count-table="items[2].countTable"
                :namespace="namespaces[2]"
                :go-ontology="items[2].ontology"
                :relative-counts="trust ? trust.totalAmountOfItems : 1"
                :loading="loading"
                :show-percentage="false">
            </go-summary>
        </template>
    </go-summary-card>
</template>

<script lang="ts">
import Vue from "vue";
import Component from "vue-class-component";
import { Prop, Watch } from "vue-property-decorator";
import { Peptide } from "./../../../business/ontology/raw/Peptide";
import { GoNamespace } from "./../../../business/ontology/functional/go/GoNamespace";
import { CountTable } from "./../../../business/counts/CountTable";
import GoDefinition, { GoCode } from "./../../../business/ontology/functional/go/GoDefinition";
import { Ontology } from "./../../../business/ontology/Ontology";
import StringUtils from "./../../../business/misc/StringUtils";
import GoProteinCountTableProcessor from "./../../../business/processors/functional/go/GoProteinCountTableProcessor";
import GoOntologyProcessor from "./../../../business/ontology/functional/go/GoOntologyProcessor";
import { FunctionalUtils } from "./../functional/FunctionalUtils";
import GoSummaryCard from "./../functional/GoSummaryCard.vue";
import GoSummary from "./../functional/GoSummary.vue";
import FunctionalTrust from "./../../../business/processors/functional/FunctionalTrust";

@Component({
    components: {
        GoSummary,
        GoSummaryCard
    }
})
export default class SingleGoSummaryCard extends Vue {
    @Prop({ required: true })
    private peptide: Peptide;
    @Prop({ required: true })
    private equateIl: boolean;

    private trust: FunctionalTrust = null;
    private trustLine: string = "";
    private loading: boolean = false;

    private namespaces: GoNamespace[] = Object.values(GoNamespace).sort();
    private items: {
        countTable: CountTable<GoCode>,
        definitions: GoDefinition[],
        title: string,
        ontology: Ontology<GoCode, GoDefinition>
    }[] = [];

    private created() {
        for (let ns of this.namespaces) {
            this.items.push({
                countTable: undefined,
                definitions: [],
                title: StringUtils.stringTitleize(ns.toString()),
                ontology: undefined
            });
        }
    }

    private mounted() {
        this.recompute();
    }

    @Watch("peptide")
    @Watch("equateIl")
    private async recompute() {
        if (this.peptide) {
            this.loading = true;

            const goProteinProcessor = new GoProteinCountTableProcessor(this.peptide, this.equateIl);

            for (let i = 0; i < this.namespaces.length; i++) {
                const namespace: GoNamespace = this.namespaces[i];
                this.items[i].countTable = await goProteinProcessor.getCountTable(namespace);

                const ontologyProcessor = new GoOntologyProcessor();
                this.items[i].ontology = await ontologyProcessor.getOntology(this.items[i].countTable);

                this.items[i].definitions.length = 0;
                this.items[i].definitions.push(
                    ...this.items[i].countTable.getOntologyIds().map(id => this.items[i].ontology.getDefinition(id))
                );
            }

            this.trust = await goProteinProcessor.getTrust();
            this.trustLine = FunctionalUtils.computeTrustLine(this.trust, "GO-term", "protein");

            this.loading = false;
        }
    }
}
</script>

<style scoped>

</style>
