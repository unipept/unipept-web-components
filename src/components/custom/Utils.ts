export default class Utils {
    /**
     * This method should be used when a specific URL should be opened in a new browser window. The method automatically
     * decides whether Electron or a default redirection should take place.
     * 
     * @param url The full url to which navigation should take place.
     */
    static openInBrowser(url: string): void {
        if (Utils.isElectron()) {
            const shell = require("electron").shell;
            shell.openExternal(url);
        } else {
            window.open(url);
        }
    }
    
    /**
     * Method that can be used to check if the app is being executed by Electron or something else.
     * 
     * @return True if the app is being executed in an Electron-powered environment.
     */
    static isElectron(): boolean {
        const userAgent = navigator.userAgent.toLowerCase();
        return userAgent.indexOf(" electron/") > -1;
    }
}
