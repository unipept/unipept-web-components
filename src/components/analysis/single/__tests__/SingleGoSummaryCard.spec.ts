import Vue from "vue"
import Vuetify from "vuetify"
import { createLocalVue, mount } from "@vue/test-utils";
import Setup from "@/test/Setup";
import { waitForCondition } from "@/test/Utils";
import SingleGoSummaryCard from "@/components/analysis/single/SingleGoSummaryCard.vue";
import DefaultCommunicationSource from "@/business/communication/source/DefaultCommunicationSource";

Vue.use(Vuetify);

const localVue = createLocalVue();

describe("SingleGoSummaryCard", () => {
    let vuetify;

    beforeEach(() => {
        vuetify = new Vuetify();
    });

    beforeAll(() => {
        const setup = new Setup();
        setup.setupAll();
    });

    it("correctly displays the correct GO-terms for a given peptide", async() => {
        const wrapper = mount(SingleGoSummaryCard, {
            localVue,
            vuetify,
            propsData: {
                peptide: "AALTER",
                equateIl: true,
                communicationSource: new DefaultCommunicationSource("http://unipept.ugent.be")
            }
        });

        await waitForCondition(() => wrapper.findAll("tr").length > 1);

        const sortButtons = wrapper.findAll("th").filter(w => w.html().includes("GO term"));
        // Filter each of the 3 tables by GO-term code.
        sortButtons.wrappers.map(w => w.trigger("click"));

        // This should include 3 containers. The first one for biological process, the second one for cellular
        // component and the last one for molecular function.
        const goContainers = wrapper.findAll(".go-table-container");

        // Make sure each of these containers has been sorted correctly before continuing
        await waitForCondition(() => goContainers.at(0).find("tbody tr").html().includes("GO:0000746"));
        await waitForCondition(() => goContainers.at(1).find("tbody tr").html().includes("GO:0005634"));
        await waitForCondition(() => goContainers.at(2).find("tbody tr").html().includes("GO:0000287"));

        // Now check for each of the tables if it contains the expected GO-terms.
        let trs = goContainers.at(0).findAll("tbody tr");
        expect(trs.at(0).html().includes("GO:0000746")).toBeTruthy();
        expect(trs.at(1).html().includes("GO:0008152")).toBeTruthy();
        expect(trs.at(2).html().includes("GO:0016567")).toBeTruthy();

        expect(goContainers.at(0).find(".v-data-footer__pagination").text()).toEqual("1-3 of 3");

        // Tests for cellular component
        trs = goContainers.at(1).findAll("tbody tr");
        expect(trs.at(0).html().includes("GO:0005634")).toBeTruthy();
        expect(trs.at(1).html().includes("GO:0005737")).toBeTruthy();
        expect(trs.at(2).html().includes("GO:0005829")).toBeTruthy();

        expect(goContainers.at(1).find(".v-data-footer__pagination").text()).toEqual("1-3 of 3");

        // Tests for molecular function
        trs = goContainers.at(2).findAll("tbody tr");
        expect(trs.at(0).html().includes("GO:0000287")).toBeTruthy();
        expect(trs.at(1).html().includes("GO:0003677")).toBeTruthy();
        expect(trs.at(2).html().includes("GO:0003678")).toBeTruthy();

        expect(goContainers.at(2).find(".v-data-footer__pagination").text()).toEqual("1-5 of 10");
    });

    it("displays the correct trust line", async() => {
        const wrapper = mount(SingleGoSummaryCard, {
            localVue,
            vuetify,
            propsData: {
                peptide: "AALTER",
                equateIl: true,
                communicationSource: new DefaultCommunicationSource("http://unipept.ugent.be")
            }
        });

        // Wait for the table to be rendered.
        await waitForCondition(() => wrapper.find(".go-trust").html().includes("protein"));

        expect(wrapper.find(".go-trust").html()).toMatchSnapshot();
    });
});
