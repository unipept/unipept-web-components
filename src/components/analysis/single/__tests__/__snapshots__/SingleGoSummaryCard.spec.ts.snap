// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleGoSummaryCard displays the correct trust line 1`] = `"<span class=\\"go-trust\\"><strong>All proteins</strong> have at least one GO-term assigned to them. </span>"`;
